### ISSUE TITLE
[BUG][MemoryLeak] Animated SKPath causes memory leak. Example Project included

### ISSUE THREAD

**Issue by AquilaAbriel** [mono/SkiaSharp#2368]
**Description**

I'm currently experimenting with animated SKPaths, but I stumbled upon a memory leak issue I couldn't resolve.
I'm not entirely sure but I suspect there is some unmanaged resource, which is not released correctly.
(Either in SKPath, or in SKSurfce, maybe related to both?)
I tried two different methods of creating an animated SKPath. Either by calling Reset() and rebuilding a new shape,
or by calling Dispose() on the old path and creating a new one.

Both options result in a memory leak.

Some further observations:
- The memory leak affects both RAM and VRAM.
- Forcing a garbage collection run does not free up the leaked memory.
- The SKPaint settings (Antialias / Stroke / Fill) don't affect the memory leak. (Apparently)
- SKCanvas matrix changes don't affect the memory leak. (Translation and Scale get applied each frame)
- Only rebuilding but not drawing the path doesn't cause a memory leak.
- Only drawing the same path, without rebuilding it, doesn't cause a memory leak. 
- Using dispose on the SKPath, instead of resetting and reusing it, doesn't stop the memory leak.
- Disposing + recreating or resetting + rebuilding the SKPath but skipping the "animation" (essentially re-adding the same points each frame) stops the memory leak?! (How is that even possible?)

After digging though some of the Google Skias documentation about SKPath, I found the [isVolatile] flag.
The description of which sounds suspiciously like what is going wrong here:

_Returns true if the path is volatile; it will not be altered or discarded by the caller after it is drawn.
[SkPath](https://api.skia.org/classSkPath.html) by default have volatile set false, allowing [SkSurface](https://api.skia.org/classSkSurface.html) to attach a cache of data which speeds repeated drawing. If true, [SkSurface](https://api.skia.org/classSkSurface.html) may not speed repeated drawing._

I don't know if the volatile setting was added to Skia in a later Version, than what SkiaSharp is build on, but both the Flag and the related setIsVolatile() function are absent in SkiaSharp.

**Code**

Alternative methods of creating the animated path (Running once each frame):
```
//-----------------------------------------------------------------------------------------------------//
//Reset and reuse existing SKPath
//-----------------------------------------------------------------------------------------------------//
PathBuffer.Reset();

//Move to first point position
PathBuffer.MoveTo(CurrentPoints[0].Position);

//Connect bezier points
for (int x = 1; x < CurrentPoints.Length; x++)
{
	PathBuffer.CubicTo(CurrentPoints[x - 1].HandleNext, CurrentPoints[x].HandleLast, CurrentPoints[x].Position);
}

//Close loop
SKPoint first = PathBuffer.GetPoint(0);
PathBuffer.CubicTo(CurrentPoints.Last().HandleNext, CurrentPoints[0].HandleLast, first);
PathBuffer.Close();
```

```
//-----------------------------------------------------------------------------------------------------//
//Dispose old SKPath and create new SKPath
//-----------------------------------------------------------------------------------------------------//
PathBuffer.Reset();
PathBuffer.Dispose();
PathBuffer = new SKPath();

//Move to first point position
PathBuffer.MoveTo(CurrentPoints[0].Position);

//Connect bezier points
for (int x = 1; x < CurrentPoints.Length; x++)
{
	PathBuffer.CubicTo(CurrentPoints[x - 1].HandleNext, CurrentPoints[x].HandleLast, CurrentPoints[x].Position);
}

//Close loop
SKPoint first = PathBuffer.GetPoint(0);
PathBuffer.CubicTo(CurrentPoints.Last().HandleNext, CurrentPoints[0].HandleLast, first);
PathBuffer.Close();
```

Drawing the animated path to the screen:
```
//Clear rendertarget
gameTarget.Clear(new SKColor(0, 0, 0, 0));

//Center canvas in the middle of the screen
gameTarget.surface.Canvas.Translate(gameTarget.width / 2f, gameTarget.height / 2f);

//Scale canvas (Path point coords are within -1 to 1 space)
gameTarget.surface.Canvas.Scale(100f, 100f);

using (SKPaint paint = new SKPaint())
{
	paint.Color = new SKColor(255, 255, 255);
	paint.IsAntialias = true;
	paint.IsStroke = true;
	paint.StrokeWidth = 0.02f;

	if (!DEBUG_SKIP_DRAW)
	{
		//Draw current path to canvas
		gameTarget.surface.Canvas.DrawPath(PathBuffer, paint);
	}
}

//Reset rendertarget matrix
gameTarget.surface.Canvas.ResetMatrix();
```

**Expected Behavior**

Correctly releasing all unmanaged data (and caches?) related to SKPath and SKSurface if SKPath gets disposed.

**Actual Behavior**

I have no idea, what is going on "behind the curtains"...
All I see is a 100mb of RAM leaking every 10 seconds. Even though I called Dispose() on everything I have access to.

**Basic Information**

- Version with issue:  2.88.3, GPU Renderer
- Last known good version:  Unknown
- IDE:  Visual Studio 2022
- Platform Target Frameworks: Windows10, .NET 7

**Screenshots**

dotMemory screenshot:
![dotMemory](https://user-images.githubusercontent.com/7701826/215282730-e7c93cf4-bc03-441b-b0b3-5474115eb346.jpg)

**Reproduction Link**

Very simplified example project to showcase the memory leak issue. (Everything importand is in the DEBUG_SKPath.cs file)
This project uses .NET 7 and requires the newest version of Visual Studio 2022 to compile:
[ExampleProjectFiles.zip](https://github.com/mono/SkiaSharp/files/10527777/ExampleProjectFiles.zip)

Memory profile produced by JetBrains dotMemory profiler (I have not enough experience to know, what I have to look for):
[MemoryProfiling.zip](https://github.com/mono/SkiaSharp/files/10527894/MemoryProfiling.zip)

**Comment by najak3d** [comment]
YES!  This is happening to us now, causing Application instability/crashing.   Our usage is less aggressive than yours, only rapid firing during user map panning where we replot the vector map lines rapidly -- using Reset() method.    Dispose() method might be better, but, still, the Reset() shouldn't be causing rampant memory leaks, and is our preference (seems more responsible than doing a rapid disposal of SKPaths).

Additionally, would love to have a means of "maintaining the previous path buffers" - more like a "SoftReset()" which maintains the buffer, and just refills it with the new data, rather than ditching/disposing of the buffer and making a new one each time.

It's a common use case for us to rapid-update the Paths... and in essence, we would like to just be "changing the content of the buffers" - not ditching/disposing the old buffer and making new ones each time.

**Comment by mattleibow** [comment]
@AquilaAbriel thanks for coming back to update the issue after over a year.

I know it is frustrating to have your issue ignored. I apologize for that. I am pretty much working alone on this in between working on other products. You are not alone in your frustrations, both other folks and myself wish I could/would do more. I am working to get more time to focus on SkiaSharp. I would be spending more weekends and evenings on this, but we just got my first baby born a few months ago so I do need to spread the love.

But, thanks again for trying/using SkiaSharp. I hope for your next C# project you will try us again.

**Comment by mattleibow** [comment]
@najak3d could you please open a new issue with your details and a repro for the issue that you are seeing.

**Comment by najak3d** [comment]
> [@najak3d](https://github.com/najak3d) could you please open a new issue with your details and a repro for the issue that you are seeing.

I was able to fully switch over to SKIA 3.116.1, and this FULLY RESOLVES the memory leak!

BUT - 3.116.1 is WORSE for us, currently, because there is a new crashing bug (even crashes it faster), associated with SKImage Disposal triggered by finalizer (might be other types of Disposal issues).  The error is:

"System.AccessViolationException: 'Attempted to read or write protected memory.  This is often an indication that other memory is corrupt.'

So we're stuck with 2.8x - until this crash is fixed (which happens consistently within a few minutes of operation).

Here are images - this one shows the Memory Leak Fixed (top=broken, vs bottom=fixed)
[](https://drive.google.com/file/d/1kClV_2w2wfnjEJJYazQIBsSN7JVY2QPh/view?usp=drive_link)

Here are the images of the new 3.116.1 crash on finalizing SKImage:
[](https://drive.google.com/file/d/1Dna_ITYcoQwIdtYc_aH_RnuQ2Bwjoab7/view?usp=drive_link)

[](https://drive.google.com/file/d/1QH76_egvSNMcBqfOLwYB52EQlnh91_2m/view?usp=drive_link)


I'm available new 24/7 to iterate with you to getting this fixed.

**Comment by najak3d** [comment]
The new bug that is preventing us from switching to Skia 3.x, is here:

[Issue 2794](https://github.com/mono/SkiaSharp/issues/2794)
